src/substrait/expression/Decimal.cpp

```diff
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "substrait/expression/Decimal.h"
+
+#include <sstream>
+
+#include "absl/numeric/int128.h"
+#include "substrait/proto/algebra.pb.h"
+
+namespace io::substrait::expression {
+
+namespace {
+
+// invertBits flips the sign of a two-complements value.
+std::string invertBits(const std::string& value) {
+  std::string newValue = value;
+  // Flip all of the bits and add one.
+  bool carryover = true;
+  for (char& b : newValue) {
+    uint8_t newB = ~(static_cast<uint8_t>(b));
+    if (carryover) {
+      newB++;
+    }
+    carryover = carryover && (newB == 0);
+    b = static_cast<char>(newB & 0xff);
+  }
+  return newValue;
+}
+
+} // namespace
+
+Decimal Decimal::fromProto(
```

Comment:
Can we rename to DecimalLiteral to keep in line with substrait proto?

Reply:
Done.